streaming/src/main/scala/org/apache/spark/streaming/util/BatchedWriteAheadLog.scala

```diff
@@ -0,0 +1,204 @@
+    active = false
+    fulfillPromises()
+    batchWriterThreadPool.shutdownNow()
+    parent.close()
+  }
+
+  /**
+   * Respond to any promises that may have been left in the queue, to unblock receivers during
+   * shutdown.
+   */
+  private def fulfillPromises(): Unit = {
+    while (!walWriteQueue.isEmpty) {
+      val RecordBuffer(_, _, promise) = walWriteQueue.poll()
+      promise.success(null)
+    }
+  }
+
+  /** Start the actual log writer on a separate thread. Visible (protected) for testing. */
+  protected def startBatchedWriterThread(): Unit = {
+    val thread = new Thread(new Runnable {
+      override def run(): Unit = {
+        while (active) {
+          try {
+            flushRecords()
+          } catch {
+            case NonFatal(e) =>
+              logWarning("Encountered exception in Batched Writer Thread.", e)
+          }
+        }
+        logInfo("Batched WAL Writer thread exiting.")
+      }
+    }, "Batched WAL Writer")
+    thread.setDaemon(true)
+    thread.start()
+  }
+
+  /** Write all the records in the buffer to the write ahead log. Visible for testing. */
+  protected def flushRecords(): Unit = {
+    try {
+      buffer.append(walWriteQueue.take())
+      val numBatched = walWriteQueue.drainTo(buffer.asJava) + 1
+      logDebug(s"Received $numBatched records from queue")
+    } catch {
+      case _: InterruptedException =>
+        logWarning("Batch Write Ahead Log Writer queue interrupted.")
+    }
+    try {
+      var segment: WriteAheadLogRecordHandle = null
+      if (buffer.length > 0) {
+        logDebug(s"Batched ${buffer.length} records for Write Ahead Log write")
+        // we take the latest record for the time to ensure that we don't clean up files earlier
+        // than the expiration date of the records
+        val time = buffer.last.time
+        segment = parent.write(aggregate(buffer), time)
+      }
+      buffer.foreach(_.promise.success(segment))
+    } catch {
+      case NonFatal(e) =>
+        logWarning(s"Batch WAL Writer failed to write $buffer", e)
+        buffer.foreach(_.promise.success(null))
+    }
+    buffer.clear()
+  }
+}
+
+/** Static methods for aggregating and de-aggregating records. */
+private[streaming] object BatchedWriteAheadLog {
+  /**
+   * Wrapper class for representing the records that we will write to the WriteAheadLog. Coupled
+   * with the timestamp for the write request of the record, and the promise that will block the
+   * write request, while a separate thread is actually performing the write.
+   */
+  private[util] case class RecordBuffer(
+      record: ByteBuffer,
+      time: Long,
+      promise: Promise[WriteAheadLogRecordHandle])
+
+  /** Aggregate multiple serialized ReceivedBlockTrackerLogEvents in a single ByteBuffer. */
+  private[streaming] def aggregate(records: Seq[RecordBuffer]): ByteBuffer = {
+    ByteBuffer.wrap(Utils.serialize[Array[Array[Byte]]](records.map(_.record.array()).toArray))
```

Comment:
The content of the ByteBuffer may be just a part of the array, e.g., this is a sliced ByteBuffer. It's better to copy the content to a new array, such as 
```
val byteArray = new Array[Byte](buffer.record.remaining())
buffer.record.get(byteArray)
byteArray
```